addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go

```diff
@@ -172,15 +186,156 @@ func addMissingFields(copyFrom, destination map[string]interface{}) error {
+						// If two slices items match(using refName to compare), we want to add missing fields from one to another.
+						// e.g., {refName: a, valuesFrom: {...}} and {refName: a} matches
+						if itemMapInFrom["refName"].(string) == itemMapInTarget["refName"].(string) {
+							foundMatch = true
+							if err := addMissingFields(itemMapInFrom, itemMapInTarget, keysToSkip); err != nil {
+								return err
+							}
+							if removeErr := removeFields(itemMapInTarget, keysToSkip); removeErr != nil {
+								return removeErr
+							}
+							updatedSlice = append(updatedSlice, itemMapInTarget)
+							break // we expect only one match
+						}
+					}
+					// If there is no match, we simply add the new slice item to the final slice
+					if !foundMatch {
+						if removeErr := removeFields(itemMapInFrom, keysToSkip); removeErr != nil {
+							return removeErr
+						}
+						updatedSlice = append(updatedSlice, itemMapInFrom)
+					}
+				}
+				destination[keyInFrom] = updatedSlice
 			}
 		}
 	}
 	return nil
 }
 
+// removeFields deletes the map entries by comparing the keys with keysToRemove
+func removeFields(m, keysToRemove map[string]interface{}) error {
+	if keysToRemove == nil || len(keysToRemove) == 0 {
+		return nil
+	}
+	for key, value := range m {
+		_, exist := keysToRemove[key]
+		if exist {
+			// Delete the field and continue to scan next key
+			delete(m, key)
+			continue
+		}
+		if value != nil && reflect.TypeOf(value).Kind() == reflect.Map {
+			// If value is a map, recursively look into it
+			if err := removeFields(value.(map[string]interface{}), keysToRemove); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// HandleExistingClusterBootstrap does the following:
+// 1. Check the annotation from the existing ClusterBootstrap CR. Add missing fields to the ClusterBootstrap CR if needed
+// 2. Clone the referenced object of ClusterBootstrapPackage.ValuesFrom into the cluster namespace. If the referenced
+//    object has embedded TypedLocalObjectReference(e.g., CPI has VSphereCPIConfig as valuesFrom, VSphereCPIConfig has
+//    a Secret reference under its Spec), this function clones the embedded object into the cluster namespace as well.
+// 3. Add OwnerReferences and Labels to the cloned objects.
+func (h *Helper) HandleExistingClusterBootstrap(clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap, cluster *clusterapiv1beta1.Cluster, tkrName, systemNamespace string) (*runtanzuv1alpha3.ClusterBootstrap, error) {
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+		clusterBootstrap.Spec.Kapp,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	if clusterBootstrap.Annotations != nil {
+		if annotationValue, annotationExist := clusterBootstrap.Annotations[constants.AddCBMissingFieldsAnnotationKey]; annotationExist {
+			tkrName = annotationValue
```

Comment:
Question: Should it be checked if this tkr actually exists?

Reply:
In few lines below, it checks the CBT (all it cares about is the presence of CBT )